Add random particle generator

gassmoeller opened the pull request:
Pretty much as the changelog entry says, this is the algorithm we use in ASPECT to generate particles according to a probability density function. The individual positions are random, but we can control the density of particles in the domain with the density function.

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     * Gassmoeller, R., Lokavarapu, H., Heien, E., Puckett, E. G., & Bangerth,
+     * W. (2018). Flexible and Scalable Particle‐in‐Cell Methods With Adaptive
+     * Mesh Refinement for Geodynamic Computations. Geochemistry, Geophysics,
+     * Geosystems, 19(9), 3596-3604.
```
You can now put this into `doc/doxygen/references.bib` and reference it with `@cite GLHPB18`.

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     * Mesh Refinement for Geodynamic Computations. Geochemistry, Geophysics,
+     * Geosystems, 19(9), 3596-3604.
+     *
+     * @param cell The cell in which a particle is generated.
```
```suggestion
     * @param[in] cell The cell in which a particle is generated.
```

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     *
+     * @param cell The cell in which a particle is generated.
+     *
+     * @param id The particle index that will be assigned to the new particle.
```
```suggestion
     * @param[in] id The particle index that will be assigned to the new particle.
```

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     *
+     * @param id The particle index that will be assigned to the new particle.
+     *
+     * @param random_number_generator A random number generator that will be
```
```suggestion
     * @param[in,out] random_number_generator A random number generator that will be
```

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     * @param random_number_generator A random number generator that will be
+     * used for the creation of th particle.
+     *
+     * @param mapping An optional mapping object that is used to map reference
```
```suggestion
     * @param[in] mapping An optional mapping object that is used to map reference
```

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+    random_particle_in_cell(
+      const typename Triangulation<dim, spacedim>::active_cell_iterator &cell,
+      const types::particle_index                                        id,
+      boost::mt19937 &              random_number_generator,
```
We can now use the C++11 class `std::mt19937`. You'll have to `#include` the file `<random>` but in return get rid of the three lines for the BOOST include.

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     *
+     * The algorithm implemented in the function is described in
+     *
+     * Gassmoeller, R., Lokavarapu, H., Heien, E., Puckett, E. G., & Bangerth,
```
same here.

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     * Mesh Refinement for Geodynamic Computations. Geochemistry, Geophysics,
+     * Geosystems, 19(9), 3596-3604.
+     *
+     * @param triangulation The triangulation associated with the @p particle_handler.
```
maybe mark up arguments as `in` and `out` as well?

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     * @param triangulation The triangulation associated with the @p particle_handler.
+     *
+     * @param probability_density_function A function with positive values
+     * (including 0) that determines the probability density of a particle
```
non-negative is the term you're looking for :-)

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -63,6 +72,99 @@ namespace Particles
+     * normalized.
+     *
+     * @param random_cell_selection A bool that determines, how the number of particles
+     * per cell is computed (see description of this function).
```
```suggestion
     * per cell is computed (see the description above).
```

bangerth commented on source/particles/generators.cc:
```diff
@@ -75,6 +142,263 @@ namespace Particles
+
+        // Calculate start id and number of local particles
+        start_particle_id =
+          llround(static_cast<double>(n_particles_to_create) *
```
```suggestion
          std::llround(static_cast<double>(n_particles_to_create) *
```
Same below.

bangerth commented on source/particles/generators.cc:
```diff
@@ -75,6 +142,263 @@ namespace Particles
+            // Uniform distribution on the interval [0,local_weight_integral).
+            // This will be used to randomly select cells for all local
+            // particles.
+            boost::random::uniform_real_distribution<double>
```
std::something, as above

bangerth commented on source/particles/generators.cc:
```diff
@@ -75,6 +142,263 @@ namespace Particles
+              if (cell->is_locally_owned())
+                {
+                  const types::particle_index particles_to_create =
+                    llround(static_cast<double>(n_local_particles) *
```
```suggestion
                    std::llround(static_cast<double>(n_local_particles) *
```

bangerth commented on source/particles/generators.cc:
```diff
@@ -75,6 +142,263 @@ namespace Particles
+                  const types::particle_index particles_to_create =
+                    llround(static_cast<double>(n_local_particles) *
+                            accumulated_cell_weights[cell_index] /
+                            local_weight_integral);
```
Doesn't this formula yield something like the *cumulative* number of particles to create, from you you get the number of particles to create by taking the difference to the previous cell's value?

bangerth commented on source/particles/generators.cc:
```diff
@@ -75,6 +142,263 @@ namespace Particles
+        // generated cell-by-cell, they will already be sorted in the correct
+        // order to be later transferred to the multimap with O(N) complexity.
+        // If we would insert them into the multimap one-by-one it would
+        // increase the complexity to O(N log(N)).
```
You made me think here. That's because the insertion of the initial few particles doesn't require log(N) steps each, but really just log(n) where n is the number of particles in the map *at the time*, not at the end! So I was wondering whether one might get something better with the naive algorithm -- which I agree we shouldn't use, but I wanted to convince myself that the complexity here is still correct.

And it is: The cost is `sum_n  log(n)`, which is approximately equal to `int_1^N  log(x) dx`, which equates to `1+N log N - N`, which is of course `O(N log N)` as claimed :-)

bangerth commented on source/particles/generators.cc:
```diff
@@ -75,6 +142,263 @@ namespace Particles
+                                            random_number_generator,
+                                            mapping);
+                  particles.insert(particles.end(),
+                                   std::make_pair(cell, particle));
```
I think you can probably use `particles.emplace_hint (particles.end(), {cell, std::move(particle)});` here.

bangerth commented:
/rebuild

gassmoeller commented:
I think I addressed all comments (some answers below). Let me know if there is anything else to-do, or if you want me to squash commits (the last one contains the changes due to your comments).

> non-negative is the term you're looking for :-)

I knew there had to be a simpler way :-)

> In other parts of the library, we have generally used vectors of size n_active_cells(), in which elements are indexed using cell->active_cell_index(). I think this is easier to understand, and it's consistent with what we do elsewhere.

I made that change here and in the parts that relied on it.

> You made me think here. That's because the insertion of the initial few particles doesn't require log(N) steps each, but really just log(n) where n is the number of particles in the map at the time, not at the end! So I was wondering whether one might get something better with the naive algorithm -- which I agree we shouldn't use, but I wanted to convince myself that the complexity here is still correct.

Yes, but it actually does not matter, since the comment was outdated, we did not actually use the vector. With the current interface of ParticleHandler I have to hand over an already constructed multimap. It should not change the complexity though, both operations should be O(N), although we have one unnecessary copy of the whole multimap happening at the moment.

bangerth commented:
Thanks. Please squash into a set of self-contained and individually complete commits (e.g., .h+.cc file, tests, changelog).

gassmoeller commented:
Done.
